### ISSUE TITLE
Hypothetical Tokamak router module

### ISSUE THREAD

**Issue by MaxDesiatov** [swiftwasm/Tokamak#203]
I had a look at [SwiftWebUI Router](https://github.com/carson-katri/swiftwebui-router) and it's great stuff, thank you for developing this @carson-katri! I hope it could be used with Tokamak with slight modification.

In the meantime, I'm just spitballing some requirements for a new router API that I hope could be designed and implemented at some point in the future:

* Ideally it would be compatible with SwiftUI, i.e. you recompile your code for Apple platforms and this router package is able to handle deep links (and maybe even user activities for Handoff too?) seamlessly.
* I'm not a fan of routes specified as strings. I wonder if an API could be created that's statically typed, but still provides enough flexibility, maybe something like this?
```swift
enum AppRoute: Codable {
  case orders
  case orderDetails(id: Int)
  case admin(AdminRoute)
}

enum AdminRoute: Codable {
  case users
  case userDetails(id: Int)
}
```
The complication here is that `Codable` doesn't play well with enums well out of the box. Maybe an alternative could be built (doesn't have to rely on `Codable` actually) with something like [CasePaths](https://github.com/pointfreeco/swift-case-paths)?
* It should be modular and composable, so that an external package (say an admin user-management package akin to the example above) could be integrated into any app and its routing tree without assuming absolute paths. E.g. I deploy the app to `example.com/app`, the app itself ideally shouldn't assume absolute paths to allow that deployment, but also if it integrates the user management package after, URLs like `example.com/app/admin/users` should work seamlessly too.
* Similarly, integrating resources should also compose without any absolute path assumptions in light of https://github.com/swiftwasm/Tokamak/pull/155 and https://github.com/swiftwasm/carton/issues/38 so that packages/targets that declare resoruces with same names don't experience URL clashes. This one's probably harder, but just as important I think.
* I'm not sure how it should integrate with the new lifecycle `App` and `Scene` types, I think routes belong to views first, but I might be wrong. Actual prototype code might be needed to clarify this.

**Comment by j-f1** [comment]
Something to consider is whether the API should be top-level (specify your routes all at once in some sort of structure inside `App` or `Scene`) or more organic (like React Router which lets you have multiple routers in different parts of the app).

**Comment by carson-katri** [comment]
I’ve actually experimented with static rendering and got something up and working fairly quickly. I can make a PR if that’d be of interest...

**Comment by j-f1** [comment]
> Static websites support? Hard to specify any requirements for that until we have static HTML rendering actually working.

It would be pretty useful if there was a way to statically export all the valid routes as HTML files/serverless functions like Gatsby or Next.js do so it could be uploaded to a static host and have proper server-side 404s.

**Comment by MaxDesiatov** [comment]
@j-f1 yes, I think I'd prefer the latter React-Router-inspired approach, otherwise I'm not sure how make relative URL composability from points 3. and 4. work.

>It would be pretty useful if there was a way to statically export all the valid routes as HTML files/serverless functions like Gatsby or Next.js do so it could be uploaded to a static host and have proper server-side 404s.

Totally, I'd imagine the renderer would traverse all views while generating HTML anyway, it could then record all the routes and split the generated HTML into separate files for separate routes at later stages of the rendering pipeline.

>I’ve actually experimented with static rendering and got something up and working fairly quickly. I can make a PR if that’d be of interest...

@carson-katri I'm very interested! I want [https://tokamak.dev](tokamak.dev) to become a proper website with docs and demos and a landing page, all built with Tokamak as much as possible, maybe we could pre-render your TokamakDocs app with that?

**Comment by j-f1** [comment]
Inspiration: [Django](https://docs.djangoproject.com/en/3.0/topics/http/urls/), [React Router](https://reactrouter.com/web/guides/quick-start), [Vue Router](https://router.vuejs.org), [SwiftUI](https://github.com/frzi/SwiftUIRouter) (react router-inspired), [Flask](https://flask.palletsprojects.com/en/1.1.x/quickstart/#routing)

**Comment by carson-katri** [comment]
I've got the code up at [`carson-katri/router`](https://github.com/carson-katri/router). There are definitely improvements to be made, so feel free to open issues/PRs there now.

**Comment by MaxDesiatov** [comment]
Just stumbled upon [this routing code](https://github.com/pointfreeco/swift-parsing/blob/346fbdcb793c4baa845ca44664ed20f719a7d83b/Sources/swift-parsing-benchmark/Routing.swift#L20) from the parsing library by the Point-Free folks, seems like an interesting alternative to `Codable`...
